html2docx.py

```diff
@@ -14,138 +15,106 @@ def __init__(self, title: str):
         super().__init__()
         self.doc = Document()
         self.doc.core_properties.title = title
-
-        self.p: Paragraph = None
-        self.runs: List[Run] = []
         self.list_style: List[str] = []
-        self.bold = 0
-        self.italic = 0
-        self.underline = 0
-        self.extra_data = ""
-
-    def feed(self, data: str) -> None:
-        super().feed(data)
-        self.finish()
+        self.href = ""
+        self._reset()
+
+    def _reset(self) -> None:
+        self.p: Optional[Paragraph] = None
+        self.r: Optional[Run] = None
+
+        # Formatting options
+        self.style = None
+        self.pre = False
+        self.attrs: List[List[str]] = []
+        self.collapse_space = True
+
+    def finish_p(self) -> None:
+        if self.r is not None:
+            self.r.text = self.r.text.rstrip()
+        self._reset()
+
+    def init_run(self, attrs: List[str]) -> None:
+        self.r = None
+        self.attrs.append(attrs)
+
+    def finish_run(self) -> None:
+        self.attrs = self.attrs[:-1]
+        self.r = None
+
+    def add_text(self, data: str) -> None:
+        if self.p is None:
+            style = self.list_style[-1] if self.list_style else self.style
+            self.p = self.doc.add_paragraph(style=style)
+        if self.r is None:
+            self.r = self.p.add_run()
+            for style_set in self.attrs:
+                for run_style in style_set:
+                    setattr(self.r.font, run_style, True)
+        self.r.add_text(data)
+
+    def add_list_style(self, name: str) -> None:
+        self.finish_p()
+        # The template included by python-docx only has 3 list styles.
+        level = min(len(self.list_style) + 1, 3)
+        suffix = f" {level}" if level > 1 else ""
+        self.list_style.append(f"{name}{suffix}")
 
     def handle_starttag(self, tag: str, attrs: List[Tuple[str, Optional[str]]]) -> None:
         if tag == "a":
-            href = next((val for name, val in attrs if name == "href"), "")
-            if href:
-                self.extra_data = " " + href
+            self.href = str(dict(attrs).get("href", ""))
```

Comment:
Avoid changing lines in a PR that’s otherwise targeted. If a line is an improvement over the previous version, the change should be suggested in a separate PR. I see `mypy` complains that the value for `href` may be `None`. This can be handled as a separate PR, to explain why the line changed instead of hiding it in a commit that’s supposed to change whitespace handling.

Besides, this line is not as clear nor as efficient as the previous version. It reads all `attrs` instead of stopping when it has found an "href", and calls `str` on `None` when the href attribute has no value, resulting in `"None"` being inserted instead of nothing.

Reply:
The reason mypy complains is actually linked to this PR. The PR introduces `href` as being an `str`. Actually, it’s an `Optional[str]`.